cmd/util/cmd/epoch-recovery/cmd/reset.go

```diff
@@ -0,0 +1,207 @@
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	"github.com/onflow/cadence"
+
+	jsoncdc "github.com/onflow/cadence/encoding/json"
+
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/engine/common/rpc/convert"
+	"github.com/onflow/flow-go/model/bootstrap"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol/inmem"
+	"github.com/onflow/flow-go/utils/io"
+)
+
+var (
+	flagBootDir string
+)
+
+// resetCmd represents a command to reset epoch data in the Epoch smart contract
+var resetCmd = &cobra.Command{
+	Use:   "reset",
+	Short: "Generates `resetEpoch` JSON transaction arguments.",
+	Long:  "Generates `resetEpoch` transaction arguments from a root protocol state snapshot and writes it to a JSON file",
+	Run:   resetRun,
+}
+
+func init() {
+	rootCmd.AddCommand(resetCmd)
+	addResetCmdFlags()
+}
+
+func addResetCmdFlags() {
+	resetCmd.Flags().StringVar(&flagBootDir, "boot-dir", "", "path to the directory containing the bootstrap files")
+	_ = resetCmd.MarkFlagRequired("boot-dir")
+}
+
+// resetRun resets epoch data in the Epoch smart contract with fields generated
+// from the root-protocol-snapshot.json
+func resetRun(cmd *cobra.Command, args []string) {
+
+	// path to the JSON encoded cadence arguments for the `resetEpoch` transaction
+	argsPath := filepath.Join(flagBootDir, "reset-epoch-args.json")
+
+	// path to the root protocol snapshot json file
+	snapshotPath := filepath.Join(flagBootDir, bootstrap.PathRootProtocolStateSnapshot)
+
+	// check if root-protocol-snapshot.json file exists under the dir provided
+	exists, err := pathExists(snapshotPath)
+	if err != nil {
+		log.Fatal().Err(err).Str("path", snapshotPath).Msgf("could not check if root protocol-snapshot.json exists")
+	}
+	if !exists {
+		log.Fatal().Str("path", snapshotPath).Msgf("root-protocol-snapshot.json file does not exists in the `boot-dir` given")
+	}
+
+	// read root protocol-snapshot.json
+	bz, err := io.ReadFile(snapshotPath)
+	if err != nil {
+		log.Fatal().Err(err).Msgf("could not read root snapshot file")
+	}
+
+	// unmarshal bytes to inmem protocol snapshot
+	snapshot, err := convert.BytesToInmemSnapshot(bz)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not convert array of bytes to snapshot")
+	}
+
+	// extract arguments from reset epoch tx from snapshot
+	cdcArgs := extractResetEpochArgs(snapshot)
+
+	// encode cadence arguments to JSON
+	encoded, err := jsoncdc.Encode(cdcArgs)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not encode cadence arguments")
+	}
+
+	// write JSON args to file
+	err = io.WriteFile(argsPath, encoded)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not write jsoncdc encoded arguments")
+	}
+}
+
+// extractResetEpochArgs extracts the required transaction arguments for the `resetEpoch` transaction
+func extractResetEpochArgs(snapshot *inmem.Snapshot) cadence.Array {
+
+	// get current epoch
+	epoch := snapshot.Epochs().Current()
+
+	// set payout
+	payout := uint64(0)
```

Comment:
Let's make this a configurable flag - I think we want the default to be `nil` cc @joshuahannan this is the `payout` parameter in `resetEpoch`

Reply:
Yep, that sounds good. Should default to `nil` unless otherwise specified I think

Reply:
`nil` being 0 right? Since the transaction doesn't accept an `Optional<UFix64>`

Reply:
This looks like it needs a change to the transaction as well -- the transaction accepts a non-optional `UFix64` so we can't actually pass in `nil`.
https://github.com/onflow/flow-core-contracts/blob/feature/epochs/transactions/epoch/admin/reset_epoch.cdc#L5